py/orbit/injection/injectparticles.py

```diff
@@ -104,54 +114,93 @@ def addParticles(self):
 
 		n_remainder = ninjected % numprocs;
 		n_inj_local = ninjected/numprocs;
-	
-		#inject the equal number of particles on each CPU
-	
+
+		#---- inject the equal number of particles on each CPU
 		i_start = rank * n_inj_local
-		i_stop = (rank+1) * n_inj_local 
-		for i in xrange(i_start, i_stop):
-			self.bunch.addParticle(x_local[i],xp_local[i],y_local[i],yp_local[i],z_local[i],dE_local[i])
-				
+		i_stop = (rank+1) * n_inj_local
+		for i in range(i_start, i_stop):
+			particleId = nPartsTotalGlobal + i
+			self.addInjectedParticle(self.bunch,x_local[i],xp_local[i],y_local[i],yp_local[i],z_local[i],dE_local[i],particleId)
+			
+		#---- inject the reminder of the particles
 		n_max_index = numprocs * n_inj_local
 
-		for i in xrange (n_remainder - 1):
-			i_cpu = int( numprocs * random.random())
+		nPartsGlobal = self.bunch.getSizeGlobal()
+		nPartsLostGlobal = self.lostbunch.getSizeGlobal()
+		nPartsTotalGlobal = nPartsGlobal + nPartsLostGlobal
+
+		for i in range(n_remainder):
+			i_cpu = random.randint(0,rank)
 			orbit_mpi.MPI_Bcast(i_cpu, mpi_datatype.MPI_INT, 0,comm)
```

Comment:
why ```random.randint(0,rank)```? shouldn't it be ```random.randint(0,numprocs)```?
Also why randomize which CPU gets a particle from remainder?
Could it be just the first `n_remainder` CPUs
```python 
i_cpu = i
i_cpu = orbit_mpi.MPI_Bcast(i_cpu, mpi_datatype.MPI_INT, 0,comm)
```

Reply:
Agree. My bad. random.randint(0,numprocs-1) it is. >Could it be just the first n_remainder CPUs. We do not want to keep overloading some first CPUs.